omnimae/vision_transformer.py

```diff
@@ -0,0 +1,1121 @@
+#!/usr/bin/env python3
+# Portions Copyright (c) Meta Platforms, Inc. and affiliates.
+# All rights reserved.
+
+# This source code is licensed under the license found in the
+# LICENSE file in the root directory of this source tree.
+
+# Code modified from 
+# https://github.com/rwightman/pytorch-image-models/blob/master/timm/models/vision_transformer.py ;
+# https://github.com/facebookresearch/deit/blob/main/models.py 
+# and https://github.com/facebookresearch/vissl/blob/main/vissl/models/trunks/vision_transformer.py
+# and is licensed under the license found in the
+# NOTICE file in the root directory of this source tree.
+
+
+# FIXME: DO NOT OPEN SOURCE THIS - we're missing proper attributions!
+
+
+import copy
+import math
+from functools import partial
+from typing import List
+
+import hydra
+import numpy as np
+import torch
+import torch.nn as nn
+import torch.utils.checkpoint as checkpoint
+from timm.models.layers import DropPath, trunc_normal_
+from torch.nn.modules.utils import _ntuple
+
+
+to_2tuple = _ntuple(2)
+
+
+def get_sinusoid_encoding_table(n_position, d_hid):
+    """Sinusoid position encoding table"""
+    # TODO: make it with torch instead of numpy
+    def get_position_angle_vec(position):
+        return [
+            position / np.power(10000, 2 * (hid_j // 2) / d_hid)
+            for hid_j in range(d_hid)
+        ]
+
+    sinusoid_table = np.array(
+        [get_position_angle_vec(pos_i) for pos_i in range(n_position)]
+    )
+    sinusoid_table[:, 0::2] = np.sin(sinusoid_table[:, 0::2])  # dim 2i
+    sinusoid_table[:, 1::2] = np.cos(sinusoid_table[:, 1::2])  # dim 2i+1
+
+    return torch.FloatTensor(sinusoid_table).unsqueeze(0)
+
+
+class PadIm2Video(torch.nn.Module):
+    def __init__(self, ntimes, pad_type, time_dim=2):
+        super().__init__()
+        self.time_dim = time_dim
+        assert ntimes > 0
+        assert pad_type in ["zero", "repeat"]
+        self.ntimes = ntimes
+        self.pad_type = pad_type
+
+    def forward(self, x):
+        if x.ndim == 4:
+            # B, C, H, W -> B, C, T, H, W
```

Comment:
Shouldn't the time axis precede the channel axis? `B, T, C, H, W` instead of `B, C, T, H, W`?

Reply:
Sorry I didn't get your question. The code just adds a new time dimension, there shouldn't be any order swapping.

Reply:
I meant why does the time axis precede the channel axis? At least that's what I understood from this comment: `# B, C, H, W -> B, C, T, H, W`.

Reply:
The time axis doesn't precede the channel axis - this function takes an image which has no time axis and adds a time axis (converts an image to a video).

That is the same setup as our omnivore paper which you can play around with in https://github.com/facebookresearch/omnivore/blob/main/inference_tutorial.ipynb.

Reply:
I understand that. I am simply asking why it's added after the channel axis. Why couldn't it be `B, T, C, H, W`?

Reply:
That's what I had mentioned earlier :) 

> Actually, we keep the time and spatial axes together so that they can be patchified together

To create a spatiotemporal patch you need the time and spatial axes to be next to each other.

Or if I'm not able to explain it I'll leave it to @kalyanvasudev to answer your question!

Reply:
Ah I now get it. Thank you for being patient. I apologize for not being attentive enough.

Reply:
No worries, glad I could help clarify things!